docs/proposals/14-wireguard-based-cluster-vpn.md

```diff
@@ -0,0 +1,178 @@
+
+We achieve this by tunneling the open vpn connection through a WireGuard tunnel, which is established from the shoot to the seed (note that WireGuard uses udp as protocol). Independent of that we can also use udp for the open vpn connection, but we can also stay with tcp as it was before. While this might look like a big change, it only introduces minor changes to the existing solution, but let's look at the details. In essence, the OpenVPN connection does not require a public endpoint in the shoot cluster but it used the internal endpoint provided by the Wireguard tunnel.
+
+The WireGuard network needs a separate network range/cidr. It has to be unique for the seed and all its shoot clusters. An example for an assumed workload of around 250 shoot clusters with enough buffer would be 192.168.0.0/22 (1024 IP addresses), i.e. 192.168.0.0-192.168.3.255. The IP addresses from this range need to be managed, but the IP address management (IPAM) using the gardener Kubernetes objects like seed and shootstate as backing store is fairly straight forward. This is especially true as we do not expect large network ranges and only infrequent IP allocations. Hence, the IP address allocation can be quite simple, i.e. scan the range for a free IP address.
+
+There is another restriction: in case shoot clusters are configured to be seed clusters this network range must not overlap with the "parent" seed cluster. If the parent seed cluster uses `192.168.0.0/22` the child seed cluster can for example use `192.168.4.0/22`. Grandchildren can however use grandparent IP address ranges.
+
+This slightly adds to the restrictions described in the current solution outline. In that the arbitrary chosen `192.168.123.0/24` range is resticted. For the purpose of this implementation we propose to extend that restriction to `192.168.128.0/17` range. Most of it would be reserved for "future use however".
+
+WireGuard can utilize the Linux kernel so that after initialization/configuration no user space processes are required. We propose to recommend the wireguard kernel module as the default solution for all seeds. For shoot clusters, the WireGuard kernel based approach is also recommended, but the user space solution should also work as we expect less traffic on the shoot side. We expect the userspace implementation to work on all operating systems supported by Gardener in case no kernel module is available.
+
+There is a user space implementation of WireGuard, which can be used on Linux distributions without the WireGuard kernel module. (WireGuard moved into the standard Linux kernel 5.6.) Our proposal can handle the kernel/user space switch transparently, i.e. we include the user space binaries and use them only when required. However, especially for the seed the kernel based solution might be more attractive. Garden Linux 318.4.0 supports WireGuard.
```

Comment:
> especially for the seed the kernel based solution might be more attractive.

Why? Can you share some statistics how it perform in the userspace and in kernelspace?

> Garden Linux 318.4.0 supports WireGuard.

Gardener is OS agnostic. This should work also on the initial seed cluster that can be GKE, AKS, EKS, Kubeadm, etc.